src/main/java/com/google/research/bleth/services/DatabaseService.java

```diff
@@ -0,0 +1,223 @@
+                Entity entity = new Entity(board.getType());
+
+                // Set properties.
+                entity.setProperty("simulationId", simulationId);
+                entity.setProperty("round", round);
+                entity.setProperty("agentId", agent.getId());
+                entity.setProperty("agentType", agent.getType());
+                entity.setProperty("rowNum", location.row);
+                entity.setProperty("colNum", location.col);
+
+                // Write to datastore.
+                datastore.put(entity);
+            }
+        }
+    }
+
+    /**
+     * Read from RealBoard and construct a Json string encoding the board state.
+     * @param simulationId is the simulation id to be retrieved.
+     * @param round is the round to be retrieved.
+     * @return a Json string encoding the board state, uses for visualization,
+     * or null if provided round exceeds simulation's maximum number of rounds.
+     */
+    public String readRealBoardState(String simulationId, int round) {
+        return readBoardState(simulationId, round, "RealBoard");
+    }
+
+    /**
+     * Read from EstimatedBoard and construct a Json string encoding the board state.
+     * @param simulationId is the simulation id to be retrieved.
+     * @param round is the round to be retrieved.
+     * @return a Json string encoding the board state, uses for visualization,
+     * or null if provided round exceeds simulation's maximum number of rounds.
+     */
+    public String readEstimatedBoardState(String simulationId, int round) {
+        return readBoardState(simulationId, round, "EstimatedBoard");
+    }
+
+    private String readBoardState(String simulationId, int round, String entityKind) {
+        // throw an exception if provided round exceeds simulation's maximum number of rounds.
+        if (!isRoundExistsInSimulation(simulationId, round)) {
+            throw new ExceedingRoundException("Provided round " + round +
+                    " exceeds maximum number of rounds of simulation " + simulationId);
+        }
+
+        // Set simple predicates.
+        Query.FilterPredicate filterBySimulationId =
+                new Query.FilterPredicate("simulationId", Query.FilterOperator.EQUAL, simulationId);
+        Query.FilterPredicate filterByRound =
+                new Query.FilterPredicate("round", Query.FilterOperator.EQUAL, round);
+
+        // Compose simple predicates.
+        Query.CompositeFilter composedFilter = Query.CompositeFilterOperator.and(filterBySimulationId, filterByRound);
+
+        // Create query and return result.
+        Query boardStateBySimulationIdAndRoundQuery = new Query(entityKind).setFilter(composedFilter);
+        PreparedQuery boardStateBySimulationIdAndRoundPreparedQuery = datastore.prepare(boardStateBySimulationIdAndRoundQuery);
+
+        // Return encoded board state.
+        return toJsonTable(boardStateBySimulationIdAndRoundPreparedQuery.asIterable(), simulationId);
```

Comment:
can this call fail?

Reply:
- `asIterable()` throws an exception if the query being executed is associated with a `Transaction` that is not active:
https://cloud.google.com/appengine/docs/standard/java/javadoc/com/google/appengine/api/datastore/PreparedQuery.html#asIterable--
This doesn't seem to be the case here since we're not using Datastore Transactions.
- `toJsonTable()` throws an exception if the simulation associated with the specified `simulationId` does not exist in datastore. But if that is the case, an exception will be thrown even before that, right when there is a call for `isRoundExistsInSimulation(simulationId, round)` (line 141).

These are all unchecked exception, therefore there is no use of the `throws` keyword in the method's signature. I will add some documentation to explain this behaviour.